File: linkedList/traversal/LLIntersection.js
import { ListNode } from '../ds.js'
/**
 * Definition of ListNode:
 * class ListNode {
 *     constructor(val = null, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

export function linked_list_intersection(head_A, head_B) {
  let ptr_A = head_A;
  let ptr_B = head_B;

  while(ptr_A != ptr_B){
    if(ptr_A){
        ptr_A = ptr_A.next;
    }
    else{
        ptr_A = head_B;
    }

    if(ptr_B){
        ptr_B = ptr_B.next;
    }
    else{
        ptr_B = head_A;
    }

    return ptr_A;
  }
}

const common = new ListNode(8);
common.next = new ListNode(10);

const head_A = new ListNode(3);
head_A.next = new ListNode(7);
head_A.next.next = common;

const head_B = new ListNode(99);
head_B.next = new ListNode(1);
head_B.next.next = common;

const intersection = linked_list_intersection(head_A, head_B);

if (intersection) {
    console.log("Intersection at node value:", intersection.val);
} else {
    console.log("No intersection");
}

File: linkedList/cpp/traversal/LLIntersection.cpp
#include<iostream>
#include<vector>

using namespace std;

struct Node{
    int data;
    Node* next;

    Node(int value){
        data = value;
        next = nullptr;
    }
};

Node* intersection(Node* head1, Node* head2){
    Node* ptr1 = head1;
    Node* ptr2 = head2;

    while(ptr1 != ptr2){
        ptr1 = ptr1 ? ptr1->next : head2;
        ptr2 = ptr2 ? ptr2->next : head1;
    }
    return ptr1;
}

int main() {

    // Common/intersection part: 7 -> 8 -> 9
    Node* common = new Node(7);
    common->next = new Node(8);
    common->next->next = new Node(9);


    // Linked List 1: 1 -> 2 -> 3 -> 7 -> 8 -> 9
    Node* head1 = new Node(1);
    head1->next = new Node(2);
    head1->next->next = new Node(3);

    // Connect List 1 to common node
    head1->next->next->next = common;


    // Linked List 2: 4 -> 5 -> 7 -> 8 -> 9
    Node* head2 = new Node(4);
    head2->next = new Node(5);

    // Connect List 2 to the SAME common node
    head2->next->next = common;
    Node* ans = intersection(head1, head2);
    cout<<ans->data;

    return 0;
}